### PR TITLE
string support

### PR DESCRIPTION
**Pull request by seibert**
This PR will add support for a wider array of string methods, as described in #3323.

Current status: Very early placeholder code using a `fake_str` subclass of string for testing to avoid clashing with current assumptions that strings are basically const tokens.  Many things still missing.

Also: A subset of this code will become the new Numba extension tutorial, as the current example (the interval class) doesn't cover enough situations for realistic extensions.

Do not use this PR for anything yet.

### CODE REVIEW

**Comment by seibert** [comment]
Also, the current implementation of find() is just a functional placeholder.  The final implementation should borrow the much more complex algorithm used by Python.

**Comment by ehsantn** [comment]
I wrote a simple benchmark below to evaluate the performance impact of checking `_kind` in `_get_code_point` during run time. On my desktop, the timer shows 811 ms with the check and 566 ms without the check which is ~40% performance impact. Thoughts?

```python
import numpy as np
import numba
from numba import types
import numba.unicode
import time
import hpat

# using HPAT since it supports time.time()
@hpat.jit
def f(str_list):
    s = 0
    cmp_str = str_list[0]
    t1 = time.time()
    for i in range(len(str_list)):
        s += int(str_list[i] == cmp_str)
    t2 = time.time()
    print("exec time", t2-t1)
    return s

n = int(1e7)
s_list = []
for i in range(n):
    s_list.append(types.fake_str(str(np.random.ranf())))

print(f(s_list))
```

**Comment by seibert** [comment]
That's pretty significant.  I had hoped the inlining of `_get_code_point` would mitigate this by allowing LLVM to move the test outside the loop, but maybe it isn't obvious that `_kind` is immutable.

**Comment by sklam** [comment]
I have looked at the `_get_code_point` speed difference.  In LLVM IR, I didn't see any special loop optimization occurred in absence of the dynamic `_kind` check.  This hints that the speed difference is due to another reason.  It turns out it is due to the refct ops surrounding `_get_code_point`.  The branching makes it difficult for the naive refct ops removal pass to delete the unnecessary refct ops.  If I force disable refct by `@njit(_nrt=False)` on the `_get_code_point`, the performance impact of the dynamic `_kind` is close to zero.

**Comment by ehsantn** [comment]
Thanks for looking at this. Vectorization isn't happening in either case?

**Comment by sklam** [comment]
As suggested by @ehsantn offline, we would want to have vectorization working here.

Things to try:

- [ ] use unsigned integer for index

**Comment by sklam** [comment]
I have started to remove the `fake_str`.  By doing so, constant unicode string usage is broken.  This requires work on the type-system to refactor the `Const` type to properly fix it.

**Comment by sklam** [comment]
Blocked by https://github.com/numba/numba/pull/3414

**Comment by sklam** [comment]
We should be cleaning up this PR to get it ready for review.

**Comment by amosbird** [comment]
Hi, the initial string proposal says 

> Given the above, bytes / Python 2 str can be represented in nopython mode as pointer and a length, Python 3 str is a pointer, a length, and an encoding enum indicating latin1, UCS-2 or UCS-4.

So is it possible to have an array of these strings? In fact I have a StringRef structure in c++ which is exactly a ptr and a length. Will it be possible to pass my structure directly to jitted functions?

**Comment by sklam** [comment]
Replying to https://github.com/numba/numba/pull/3363#issuecomment-428802190 and https://github.com/numba/numba/pull/3363#issuecomment-429086040.

https://github.com/numba/numba/pull/3363/commits/32b358519749f44be526aebffd65d04d309f6d06 added performance fix by disable refcount ops in critical loop.

The reason that `_cmp_region` can't be vectorized because of the early escape in it (https://github.com/numba/numba/pull/3363/files#diff-a21787256ad779b4c4c5c99915c71025R264).  Removing the escaping branch there is needed to enable vectorization but it will add redundant character comparison.  It will add additional work and high code complexity.

**Comment by seibert** [comment]
We could partially unroll that comparison loop?

**Comment by seibert** [comment]
(due to Python strings (and Numba-allocated ones) having a null code point at the end, it would be safe to unroll the loop by 2 without special handling at the end.)

File: numba/tests/test_unicode.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+    def test_concat(self, flags=no_pyobj_flags):
+        pyfunc = concat_usecase
+        cfunc = njit(pyfunc)
+        for a in UNICODE_EXAMPLES:
+            for b in UNICODE_EXAMPLES[::-1]:
+                self.assertEqual(pyfunc(a, b),
+                                 cfunc(a, b),
+                                 "'%s' + '%s'?" % (a, b))
+
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: This needs test cases that trigger `IndexError: string index out of range`

File: numba/tests/test_unicode.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: These tests all fail:
```python
    def test_pointless_slice(self, flags=no_pyobj_flags):
        def pyfunc(a):
            return a[:]
        cfunc = njit(pyfunc)
        args = ['a']
        self.assertEqual(pyfunc(*args), cfunc(*args))

    def test_walk_backwards(self, flags=no_pyobj_flags):
        def pyfunc(a):
            return a[::-1]
        cfunc = njit(pyfunc)
        args = ['a']
        self.assertEqual(pyfunc(*args), cfunc(*args))

    def test_stride_slice(self, flags=no_pyobj_flags):
        def pyfunc(a):
            return a[::2]
        cfunc = njit(pyfunc)
        args = ['a']
        self.assertEqual(pyfunc(*args), cfunc(*args))
```

File: numba/pythonapi.py [hunk 3]
```diff
@@ -14,6 +14,13 @@
 from numba import types, utils, cgutils, lowering, _helperlib
 
 
+if PYVERSION >= (3,3):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/pythonapi.py]: Imported in `unicode.py` and used in `_get_code_point`

File: numba/pythonapi.py [hunk 4]
```diff
@@ -1117,6 +1129,31 @@ def string_as_string_and_size(self, strobj):
...
+        """
+        Returns a tuple of ``(ok, buffer, length, kind)``.
+        The ``ok`` is i1 value that is set if ok.
+        The ``buffer`` is a i8* of the output buffer.
+        The ``length`` is a i32/i64 (py_ssize_t) of the length of the buffer.
+        The ``kind`` is a i64 (int64) of the Unicode kind constant
+        """
+        if PYVERSION >= (3, 3):
+            p_length = cgutils.alloca_once(self.builder, self.py_ssize_t)
+            p_kind = cgutils.alloca_once(self.builder, Type.int())
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/pythonapi.py]: The docstring is outdated.  `kind` should be i32.

File: numba/pythonapi.py [hunk 5]
```diff
@@ -1117,6 +1129,31 @@ def string_as_string_and_size(self, strobj):
 
         return (ok, buffer, self.builder.load(p_length))
 
+    def string_as_string_size_and_kind(self, strobj):
+        """
+        Returns a tuple of ``(ok, buffer, length, kind)``.
+        The ``ok`` is i1 value that is set if ok.
+        The ``buffer`` is a i8* of the output buffer.
+        The ``length`` is a i32/i64 (py_ssize_t) of the length of the buffer.
+        The ``kind`` is a i64 (int64) of the Unicode kind constant
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/pythonapi.py]: I'll be updating the doc.

File: numba/pythonapi.py [hunk 6]
```diff
@@ -1135,6 +1172,13 @@ def string_from_string(self, string):
         fn = self._get_function(fnty, name=fname)
         return self.builder.call(fn, [string])
 
+    def string_from_kind_and_data(self, kind, string, size):
+        fnty = Type.function(self.pyobj, [Type.int(), self.cstring, self.py_ssize_t])
+        if PYVERSION >= (3, 3):
+            fname = "PyUnicode_FromKindAndData"
+            fn = self._get_function(fnty, name=fname)
+        return self.builder.call(fn, [kind, string, size])
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/pythonapi.py]: good catch.

File: numba/tests/test_unicode.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    'ascih',
+    'ascii',
+    'ascij',
+    '大处着眼，小处着手',
+    '大处着眼，小处着手。',
+    '大处着眼，小处着手。🐍⚡',
+]
+
+
+# FIXME
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_unicode.py]: ping @seibert

File: numba/typing/unicode_str.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import math
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/typing/unicode_str.py]: no longer needed

File: numba/tests/test_unicode.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (seibert) [reply numba/tests/test_unicode.py]: this and the following two lines:
```
UNICODE_EXAMPLES = [x for x in UNICODE_EXAMPLES]
UNICODE_ORDERING_EXAMPLES = [x for x in UNICODE_ORDERING_EXAMPLES]
```
should be removed.  They originally were converting the test strings to the fake string type for development, but now that is no longer needed.

File: numba/tests/test_unicode.py (hunk 2 above)
Comment (sklam) [reply numba/tests/test_unicode.py]: these tests are added

File: numba/tests/test_unicode.py (hunk 1 above)
Comment (sklam) [reply numba/tests/test_unicode.py]: error tests added

File: numba/unicode.py [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+    #     accept ch.  Will truncate ch to make it fit.
+    #   * Assumes that i is within the valid boundaries of the function
+    if a._kind == PY_UNICODE_1BYTE_KIND:
+        set_uint8(a._data, i, ch)
+    elif a._kind == PY_UNICODE_2BYTE_KIND:
+        set_uint16(a._data, i, ch)
+    elif a._kind == PY_UNICODE_4BYTE_KIND:
+        set_uint32(a._data, i, ch)
+    else:
+        pass    # FIXME: wchar?
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/unicode.py]: turned into a raise

File: numba/unicode.py [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+        if a_chr < b_chr:
+            return -1
+        elif a_chr > b_chr:
+            return 1
+
+    return 0
+
+
+@njit
+def _cmp(a, b):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/unicode.py]: removed

File: numba/unicode.py [hunk 11]
```diff
@@ -246,8 +249,10 @@ def _kind_to_byte_width(kind):
         return 2
     elif kind == PY_UNICODE_4BYTE_KIND:
         return 4
+    elif kind == PY_UNICODE_WCHAR_KIND:
+        raise AssertionError("PY_UNICODE_WCHAR_KIND string encountered")
     else:
-        return 4    # FIXME: wchar
+        raise AssertionError("Unexpected unicode encoding encounter")
```
Comment (stuartarchibald) [inline numba/unicode.py]: s/encounter/encountered/ ? however, as this is unreachable, it can probably be left!

Comment (seibert) [reply numba/unicode.py]: fixed

**Comment by stuartarchibald** [comment]
Thanks for the fixes, great job @seibert @sklam, merging!

**Comment by dlee992** [comment]
> I have looked at the `_get_code_point` speed difference. In LLVM IR, I didn't see any special loop optimization occurred in absence of the dynamic `_kind` check. This hints that the speed difference is due to another reason. It turns out it is due to the refct ops surrounding `_get_code_point`. The branching makes it difficult for the naive refct ops removal pass to delete the unnecessary refct ops. If I force disable refct by `@njit(_nrt=False)` on the `_get_code_point`, the performance impact of the dynamic `_kind` is close to zero.

Hi, @graham. As siu implied several years ago, `_nrt` could affect refct ops presence, rather than not doing any optimization!
